graspy/subgraphing/sg.py

```diff
@@ -0,0 +1,146 @@
+import numpy as np
+from scipy.stats import fisher_exact
+
+
+def construct_contingency(graphs, labels):
+    """
+    Construct an array of edgewise contingency matrices for a set of labeled graph samples.
+    
+    Parameters
+    ----------
+    graphs: array-like, shape (n_vertices, n_vertices, s_samples)
+        A set of labeled (n_vertices, n_vertices) unweighted graph samples.
+        
+    labels: vector, length (s_samples)
+        A vector of class labels. There must be a maximum of two classes.
+        
+    Returns
+    -------
+    out: array-like, shape (n_vertices, n_vertices, 2, 2)
+        An array that stores a 2-by-2 contingency matrix for each edge.
+    """
+
+    if not isinstance(graphs, np.ndarray):
+        msg = "Input array 'graphs' must be np.ndarray, not {}.".format(type(graphs))
+        raise TypeError(msg)
+    if not isinstance(labels, (list, np.ndarray)):
+        msg = "Input vector 'labels' must be list or np.ndarray, not {}.".format(
+            type(labels)
+        )
+        raise TypeError(msg)
+
+    shape = np.shape(graphs)
+    if len(shape) != 3:
+        msg = "Input array 'graphs' must be 3-dimensional with shape (n_vertices, n_vertices, s_samples)."
+        raise ValueError(msg)
+    if shape[0] != shape[1]:
+        msg = "Input array 'graphs' must have matching number of vertices."
+        raise ValueError(msg)
+
+    if len(np.shape(labels)) != 1:
+        msg = "Input vector 'labels' must be 1-dimensional."
+        raise ValueError(msg)
+    if len(np.unique(labels)) > 2:
+        msg = "Input arrays must have a maximum of two classes, not {}.".format(
+            len(np.unique(labels))
+        )
+        raise ValueError(msg)
+    if len(labels) != shape[2]:
+        msg = "Input vector length must match the number of graph samples."
+        raise ValueError(msg)
+
+    nverts = np.shape(graphs)[0]
+    out = np.zeros((nverts, nverts, 2, 2))
+    rowsum1 = sum(labels)
+    rowsum0 = len(labels) - rowsum1
+    for i in range(nverts):
+        for j in range(nverts):
+            a = sum(graphs[i, j, labels == 0])
+            b = sum(graphs[i, j, :]) - a
+            out[i, j, :, :] = [[a, rowsum0 - a], [b, rowsum1 - b]]
+    return out
+
+
+def estimate_signal_subgraph(graphs, labels, constraints):
```

Comment:
Turn this into a class instead of a function

Reply:
Should I leave the contingency matrix function as is then?

Reply:
Yea. It can just be a private function in your class

Reply:
Cool. I'll get started now

Reply:
So at the moment, I only have the signal subgraph estimator and the contingency matrix constructor. Would it be okay if I just had

self.graphs and self.labels set through __init__ method
self.subgraph set through a transform method
And the contingency constructor through a private function, 
without having a normal fit method? Or is there a particular reason to have both a fit and fit_transform method?

Reply:
Alternatively I can have the computation of the signal subgraph done through fit, and just return the actual subgraph through the transform method, which is what Ali did for faq